### ISSUE TITLE
Rule for platform-independent line separator needed. 

### ISSUE THREAD

**Issue by Fernal73** [checkstyle/checkstyle#7026]
Java is a platform-independent language.

Checkstyle has no rule(s) that enforces platform-independent line separator.

Code is cleaner if 
`System.lineSeparator()` ,`%n` in `System.out.printf(...)` or simply `System.out.println(...)`
are used instead.

Other than unwanted String objects created by concatenation and for file formats in editors to remain consistent across platforms and behave exactly the same, I see no reason why such a rule should not be formulated. 

Such a rule would be applicable across an application.

If you really need to have only one line separator across platforms, the system property can always be overridden. 

https://stackoverflow.com/questions/22681534/java-change-system-new-line-character

Referring this [discussion](https://groups.google.com/forum/m/#!msg/checkstyle/aIiTeiYCa1I/xAMlr4Z4AwAJ)  in the Checkstyle Google group. 

--------------
Module definition:
```xml
<module name="PlatformIndependentLineSeparator">
<property name="separators" value="CR, CRLF, LF" />
<\module>
```
I'm assuming the CR, CRLF and LF can be mapped to String constants in the code. If not, the actual values can be substituted and a check implemented so that malicious values are not accepted. This should come under Checker module, not TreeWalker. 

Example code:

```java
package cstests;

/***
This class lists the varied combinations where the line separator denoted by \n
is to be replaced by either a call to System.lineSeparator(), %n in a
System.out.printf() or System.getProperty("line.separator").


The same rule should apply to \r\n as well, if specified.
\n and \r\n must not be flagged in this comment block.
***/

public class TestPlatformIndependentLineSeparator {

  /***
  This class lists the varied combinations where the line separator denoted by
  \n is to be replaced by either a call to System.lineSeparator(), %n in a
  System.out.printf() or System.getProperty("line.separator").


  The same rule should apply to \r\n as well, if specified.
  \n and \r\n must not be flagged in this comment block either.
  ***/

  // test if \n is flagged here. It shouldn't.
  // test if \r\n is flagged here. It shouldn't.

  private static final String testString = "\nTest to catch line separator\n";

  private static final String testString2 =
      "\r\nTest to catch line separator\r\n";
  
  private static final String testString3 = 
    "\\nTest to escape line separator\\n";

  private static final String testString4 =
      "\\r\\nTest to escape line separator\\r\\n";

  // this must be flagged
  private static final String CRLF = "\r\n";

  // this must be flagged
  private static final String LF = "\n";

  static {
    System.setProperty("line.separator","\r\n");
    // above must not be flagged

    System.out.println(testString);
    System.out.println(testString2);
    System.out.println(testString3);
    System.out.println(testString4);
    StringBuilder sb = new StringBuilder();
    sb.append('\n');  // this should be flagged
    sb.append("\r\n"); // this should be flagged
    System.out.println(sb.toString());
    System.out.printf("%s\n", sb.toString());
    // the above line must be flagged

    char[] chars = new char[] {'\n', '\r', 'a', 'b', 'c'};
    // the above shouldn't be flagged. Maybe, it // should or someone could do
    // the following

    String charString = new String(chars);
    System.out.print(charString);

    StringBuffer sb1 = new StringBuffer();
    sb1.append('\n');  // this should be flagged
    sb1.append("\r\n"); // this should be flagged
    System.out.println(sb1.toString());
    System.out.printf("%s\n", sb1.toString());
    // the above line must be flagged
    
    System.setProperty("line.separator",LF);
  }
  
  public static void main(String... args) {
    System.setProperty("line.separator","\r\n");
    // above must not be flagged

    System.out.println(testString);
    System.out.println(testString2);
    System.out.println(testString3);
    System.out.println(testString4);
    StringBuilder sb = new StringBuilder();
    sb.append('\n');  // this should be flagged
    sb.append("\r\n"); // this should be flagged
    System.out.println(sb.toString());
    System.out.printf("%s\n", sb.toString());
    // the above line must be flagged

    char[] chars = new char[] {'\n', '\r', 'a', 'b', 'c'};
    // the above shouldn't be flagged. Maybe, it // should or someone could do
    // the following

    String charString = new String(chars);
    System.out.print(charString);

    StringBuffer sb1 = new StringBuffer();
    sb1.append('\n');  // this should be flagged
    sb1.append("\r\n"); // this should be flagged
    System.out.println(sb1.toString());
    System.out.printf("%s\n", sb1.toString());
    // the above line must be flagged
   
    // This line is not to be flagged
    System.out.printf("%s%n", sb1.toString());
    System.setProperty("line.separator",LF);
  }
}
```

EDITED FOR EXAMPLES ABOVE. 

The code comments state which code statements are to be flagged and which aren't. No comment implies no flagging.

There could be more statements that use the PrintStream, Writer and its subclasses and their methods. But since, this check will simply check for '\n', '\r' and '\r\n', there's no real need to add these examples.

**Comment by Fernal73** [comment]
Do you still need me to design this rule when I won't be implementing it? Unfortunately, that I won't be able to. I have issues that prevent me doing so. I can try and provide you the rest.

**Comment by romani** [comment]
> Other open source projects such as PMD don't need the raisers of issues to fill in the design or implementation details themselves.

in my long experience, if there is just rough idea, it never end up in implementation, as we have 500 other issues, with more clear definitions of what is required to do. Outside contributors also do not pick up uncertain/unclear issues. So just shooting of idea does not worth that much. So by demand for details to help your idea to not fall in black hole of open issues.

**Comment by Fernal73** [comment]
I think it's doable after checking out the Checkstyle API especially if comments are ignored. Unfortunately, this will have to disappear into the black hole, as you said, of open issues. I simply don't have the bandwidth for anything of the sort.

**Comment by Fernal73** [comment]
I'm not on it. Unfortunately , I can't get anyone else interested in implementing it, either.

**Comment by romani** [comment]
If somebody come up with more exact design proposal, please fill free to open new issue

**Comment by Fernal73** [comment]
What do you mean by a more exact design proposal? Are you unhappy with the module or the examples?

I'm inclined to go with a much simpler design where strings are checked for '\n','\r' and/or '\r\n'. These ought to cover all three platforms, *nix, windows and apple. 

If there are exceptions that need to be allowed, they can be suppressed. Do you envisage a pollution of the code with too many suppressions of false positives via the design proposed above.

I envisage the use of System.setProperty("line.separator",..) in tandem with this rule if developers insist on platform-independent line separator. 

Some Checkstyle rules have CRLF, CR and LF already defined. Is there a clash with these definitions?

**Comment by Fernal73** [comment]
The title of this issue needs to be changed to platform-dependent line.separator.

The application programmer will have to use his or her own line separator if he or she wishes to use a platform-independent one.

**Comment by Fernal73** [comment]
"FS: Format string should use %n rather than n (VA_FORMAT_STRING_USES_NEWLINE)
This format string includes a newline character (\n). In format strings, it is generally preferable to use %n, which will produce the platform-specific line separator."

https://spotbugs.readthedocs.io/en/stable/bugDescriptions.html

FindBugs (now SpotBugs) has a rule that enforces replacement of '\n' by platform-dependent line-separator.

Developers might not wish to use my rule since it makes the code verbose when either lineSeparator()  or System.getProperty() are used instead. 

Any ideas?

**Comment by Fernal73** [comment]
The rule is looking a lot more now. See comments above.
